### PR TITLE
Add required file content description field. Fixes #542.

### LINKED ISSUE #542: Metadata to describe the content of a File.
**For which schema is a change/update being suggested?**

~`type/file/supplementary_file.json` and~ `core/file/file_core.json`

**What should the change/update be?**

For `file_core.json`:
  - [x] add a field that describes the content of the file (`content_description`)
  - [x] add either ontology to describe valid values for this field (EDAM, EDAM-BIOIMAGING)

For `supplementary_file.json`:
  - [x] ~remove `file_description` which is now redundant with `content_description`~ - Moving to a new ticket as this is a major version change and we are in a metadata freeze at the moment.

**What new field(s) need to be changed/added?**

* Field name: content_description
* Field description: General description of the contents of the file.
* Field type: string
* Required: yes(?)
* Example: protocol; read alignment; reference gene list; matrix; ...
* CV or enum: Should be an ontology (EDAM, EDAM-BIOIMAGING)

For full discussion (requirements and options) please see the [RFC](https://docs.google.com/document/d/1Gy81_EnS3-pM8nHSUPhHXwHn-jp-EXE2oDR1lwQzfy4/edit#heading=h.6c6c1iwxunp).

Useful bit: 

> Add a ‘file content type’ ontology field onto the supplementary file entity (this is not the same as file format, this would be about information contained in the file). Previously we discussed typing the supplementary files themselves rather than the process/protocol but the field would be the same. Either a granular enum requiring updates as we are presented with new files or a more general typing. The discussion on the more general typing suggested this was of limited use. A protocol would not be required to determine what the supplementary file contained.

**Why is the change requested?**

Our current mechanism for describing a file (sequencing, analysis, supplementary, imaging) does not capture what it contains which is useful for downstream interpretation. The Browser could use this information to display to users what is actually in the files they are about to download. Currently, only the file extension is shown, but this is not sufficient information to help users decide if they are downloading the files they need/want. Showing the file name is also not sufficient.

We should get a full list of file contents uploaded by data contributors and produced by the analysis pipelines. These values should populate the enum and can be extended when the DCP encounters files with new content (e.g. imaging).

#### New terms

- **Protocol** - A procedure for carrying out a scientific experiment ([citation](https://en.oxforddictionaries.com/definition/protocol)). - **EDAM term under Data** - [REQUESTED](https://github.com/HumanCellAtlas/ontology/issues/26)
- **Starfish File** - for tiffs, zarrays and all data json related to a starfish formatted project. Not including the key file experiment.json that requires explicit tagging for the validator to work. **missing from EDAM ontology** - ON HOLD UNTIL MORE CLARITY
- **Starfish Experiment File** - marker for a key file experiment.json. This file points to all other starfish files and is required for validation. **missing from EDAM ontology** - ON HOLD UNTIL MORE CLARITY

#### Existing terms

- [DNA sequence (raw)](https://www.ebi.ac.uk/ols/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fdata_3497) or [Sequence](https://www.ebi.ac.uk/ols/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fdata_2044) - for describing raw fastq files
- [Sample annotation](https://www.ebi.ac.uk/ols/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fdata_3113) or [Reference sample report](https://www.ebi.ac.uk/ols/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fdata_3567) - for describing extra sample or donor information
- [Electronic health record](https://www.ebi.ac.uk/ols/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fdata_3861) - for describing clinical information about donors
- [Experimental report](https://www.ebi.ac.uk/ols/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fdata_2531) - for describing any non-protocol, non-sample information about the experiment that generated data
- ~[Codebook]() -Extracted in metadata so explicit tag **not needed**. This is pointed to from experiment.json so explicit tag is not needed for validators to run.~

#### New terms

- **Alignment reference** - One or a set of reference molecular sequences, structures, or profiles used for alignment of genomic or transcriptomic experimental data. - **EDAM term under Data/Alignment** - [REQUESTED](https://github.com/HumanCellAtlas/ontology/issues/25)
- **Count matrix** - A table of values representing summarised read counts per genomic region (e.g. gene, transcript). ([modified citation](https://www.ebi.ac.uk/training/online/glossary/normalised-read-count-matrix)) - **EDAM term under Matrix** - [REQUESTED](https://github.com/HumanCellAtlas/ontology/issues/23)
- **Quality control metrics** - Results of testing an output or a sample of an output against the desired specification or standard ([modified from citation](https://en.oxforddictionaries.com/definition/quality_control)). - **EDAM term under Report** - [REQUESTED](https://github.com/HumanCellAtlas/ontology/issues/24)

#### Existing terms

- [Sequence alignment](https://www.ebi.ac.uk/ols/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fdata_0863) - for describing bam/bai output files
- ~[Genome index](https://www.ebi.ac.uk/ols/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fdata_3210) - for describing what?~

#### Existing terms

- [Gene expression matrix](https://www.ebi.ac.uk/ols/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fdata_3112) - for describing gene expression matrices (do we need this?) This may be in zarr format at the moment. This is different to zarr formatted image files generated by starfish.
- [Sequence metadata](https://www.ebi.ac.uk/ols/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fdata_3732) - for describing a downloaded, merged metadata file about sequencing data
- [Image metadata](https://www.ebi.ac.uk/ols/ontologies/edam/terms?iri=http%3A%2F%2Fedamontology.org%2Fdata_3546) - for describing a downloaded, merged metadata file about imaging data

### PR DESCRIPTION
**Pull request by malloryfreeberg**
### Release notes

For file_core:
- Added field content_description. Fixes #542.
- Added new file_content_ontology schema. Fixes #542.

### Reviews
- Needs 2 Reviewers because this is a minor version change AND a new schema
- Second Reviewer to merge changes after **Monday June 17** after HCA OLS is released.

### CODE REVIEW

File: json_schema/module/ontology/file_content_ontology.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            "description": "General description of the contents of the file.",
+            "type": "string",
+            "example": "DNA sequence (raw); Sequence alignment",
+            "user_friendly": "Content description"
+        },
+        "ontology": {
+            "description": "An ontology term identifier in the form prefix:accession.",
+            "type": "string",
+            "graph_restriction":  {
+                "ontologies" : ["obo:DATA"],
```
Comment (daniwelter) [inline json_schema/module/ontology/file_content_ontology.json]: This should be obo:edam rather than obo:DATA - DATA is a subsetting strategy, not to the ontology namespace

File: tests/schema_test_files/process/test_pass_analysis_process.json [hunk 2]
```diff
@@ -129,6 +129,5 @@
     }
   ],
   "timestamp_start_utc": "2018-03-22T16:19:51.600Z",
-  "timestamp_stop_utc": "2018-03-22T16:20:25.621Z",
-  "schema_type": "process"
```
Comment (daniwelter) [inline tests/schema_test_files/process/test_pass_analysis_process.json]: How is this change related to the overall PR?

**Comment by daniwelter** [comment]
I would prefer for this PR to be put on hold until we have a clear timeframe for the inclusion and release of new EDAM terms

File: tests/schema_test_files/process/test_pass_analysis_process.json (hunk 2 above)
Comment (malloryfreeberg) [reply tests/schema_test_files/process/test_pass_analysis_process.json]: Hmm. This might have been a product of me having to resolve conflicts, given that this PR has been open for a while. I'll remove the change.

Comment (malloryfreeberg) [reply tests/schema_test_files/process/test_pass_analysis_process.json]: Oh, I see what happened. The line `"schema_type": "process"` is repeated in line 80 of the file. So, at some point it must have been duplicated. Are you ok with keeping the removal here, such that the test JSON only has 1 line of `"schema_type": "process"`? It's interesting that this didn't fail validation... @daniwelter

File: json_schema/module/ontology/file_content_ontology.json (hunk 1 above)
Comment (malloryfreeberg) [reply json_schema/module/ontology/file_content_ontology.json]: Thanks. Is the next line, `"classes": ["DATA:0006"],`, still correct?

**Comment by zperova** [comment]
I agree with @daniwelter

**Comment by malloryfreeberg** [comment]
@daniwelter I understand wanting to wait until the new EDAM terms are released. But I'm also worried about this change being a blocker for the data browser and Azul in terms of being able to develop new features related to how files are displayed. With this being an optional field, it means that we don't actually have to use it right now in real data, but we could make some example JSON documents using the new field - at least with the `.text` field filled in - even if the correct mapped ontology terms don't exist yet in EDAM. Does it makes sense to release the schema change now, and then when the EDAM terms come through, everything is all ready to go?

File: tests/schema_test_files/process/test_pass_analysis_process.json (hunk 2 above)
Comment (daniwelter) [reply tests/schema_test_files/process/test_pass_analysis_process.json]: Happy for this to stay in, just wanted to make sure it hadn't snuck in completely by accident.

I don't think validation would pick up the same field being present twice. Might be worth testing in a number of scenarios - or make it another use case for the linter.

File: json_schema/module/ontology/file_content_ontology.json [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        },
+        "ontology": {
+            "description": "An ontology term identifier in the form prefix:accession.",
+            "type": "string",
+            "graph_restriction":  {
+                "ontologies" : ["obo:edam"],
+                "classes": ["DATA:0006"],
+                "relations": ["rdfs:subClassOf"],
+                "direct": false,
+                "include_self": true
```
Comment (daniwelter) [inline json_schema/module/ontology/file_content_ontology.json]: Please set include_self to false as DATA:0006 is a place holder that should never be used for annotation

File: json_schema/module/ontology/file_content_ontology.json (hunk 1 above)
Comment (daniwelter) [reply json_schema/module/ontology/file_content_ontology.json]: I have separate issues with the use of DATA:0006 as the root of this restriction (see latest comment on the ontology file changes), but no practical alternative apart from listing each category separately.
So yes, the class restriction is still correct.

File: json_schema/module/ontology/file_content_ontology.json (hunk 3 above)
Comment (malloryfreeberg) [reply json_schema/module/ontology/file_content_ontology.json]: done

**Comment by malloryfreeberg** [comment]
Ok, let's plan to merge this in after the release on June 17. I'll update the main ticket.

**Comment by malloryfreeberg** [comment]
@zperova if you could review/merge, that would be great!
